### PR TITLE
fix scene clipping and the vertical translation bug

### PR DESCRIPTION
**Pull request by jkrumbiegel**
this already makes makielayout work much better than before

### CODE REVIEW

**Comment by asinghvi17** [comment]
Thanks a lot - should I merge now, or is there more you wanted to add?

(p.s. you should be able to create branches in the main repo now)

File: src/CairoMakie.jl [hunk 1]
```diff
@@ -531,15 +531,41 @@ function draw_plot(scene::Scene, screen::CairoScreen, primitive::Combined)
 end
 
 function draw_plot(screen::CairoScreen, scene::Scene)
+
+    # get the parent area to correct for its pixel size when translating
+    parent_area = if isnothing(scene.parent)
```
Comment (SimonDanisch) [inline src/CairoMakie.jl]: There is: https://github.com/JuliaPlots/AbstractPlotting.jl/blob/master/src/scenes.jl#L199

File: src/CairoMakie.jl [hunk 2]
```diff
@@ -531,15 +531,41 @@ function draw_plot(scene::Scene, screen::CairoScreen, primitive::Combined)
...
+        pixelarea(scene)[]
+    else
+        pixelarea(scene.parent)[]
+    end
```
Comment (asinghvi17) [inline src/CairoMakie.jl]:
```suggestion
    parent_area = pixelarea(root(scene))[]
```
if you want the full thing

Comment (jkrumbiegel) [reply src/CairoMakie.jl]: not sure, do we need the parent or the root? This would only matter for scenes that are child scenes of child scenes

Comment (asinghvi17) [reply src/CairoMakie.jl]: right, but we should at least handle it (consider an animation of multiple scenes with a popup)

Comment (asinghvi17) [reply src/CairoMakie.jl]: plus, if Cairo is using root-level coordinates, we would need to as well

**Comment by asinghvi17** [comment]
Fixes #27

File: src/CairoMakie.jl (hunk 2 above)
Comment (SimonDanisch) [reply src/CairoMakie.jl]: Hm, although, we should just use the rect of the cairoscreen, no? Thats what we're trying to draw after all

Comment (asinghvi17) [reply src/CairoMakie.jl]: Good point - we can access that by:
```suggestion
    parent_area = round.(Int, (screen.surface.width, screen.surface.height)
```

Comment (asinghvi17) [reply src/CairoMakie.jl]: actually, scrap this

File: src/CairoMakie.jl [hunk 3]
```diff
@@ -531,15 +531,41 @@ function draw_plot(scene::Scene, screen::CairoScreen, primitive::Combined)
...
+    parent_area_height = widths(parent_area)[2]
```
Comment (asinghvi17) [inline src/CairoMakie.jl]:
```suggestion
parent_area_height = round(Int, screen.surface.height)

**Comment by asinghvi17** [comment]
There seems to be a heatmap bug of some kind:
```julia
scene = Scene(resolution=(800,600), camera=campixel!);
layout = GridLayout(scene, alignmode=Outside(20));
la = layout[1,1] = LAxis(scene)
heatmap!(la, rand(100,100))
scene
```
iTerm2 p9AaEO

**Comment by jkrumbiegel** [comment]
that looks exactly like the old behavior, sure you're on the right branch?

**Comment by asinghvi17** [comment]
dang, my branches are everywhere today, sorry about that...

File: src/CairoMakie.jl (hunk 3 above)
Comment (jkrumbiegel) [reply src/CairoMakie.jl]: why should the parent always be the screen? if we have multiple nested scenes?

**Comment by jkrumbiegel** [comment]
I tested this with a child scene of a child scene and the newest commit gave the correct placement of bounding boxes. In the previous version, the grandchild scene was misplaced. So I think this can be merged.

```julia
scene = Scene(camera = campixel!)
subscene = Scene(scene, BBox(100, 500, 100, 500), camera = campixel!)
subscene2 = Scene(subscene, BBox(200, 300, 200, 300), camera = campixel!)

poly!(subscene, AbstractPlotting.zero_origin(pixelarea(subscene)[]), color = :transparent, strokecolor = :black, strokewidth = 1, show_axis = false)
poly!(scene, pixelarea(subscene)[], color = :transparent, strokecolor = :red, strokewidth = 1, show_axis = false)

poly!(subscene2, AbstractPlotting.zero_origin(pixelarea(subscene2)[]), color = :transparent, strokecolor = :black, strokewidth = 1, show_axis = false)
poly!(scene, pixelarea(subscene2)[], color = :transparent, strokecolor = :red, strokewidth = 1, show_axis = false)
scene
```

grafik

File: src/CairoMakie.jl (hunk 3 above)
Comment (asinghvi17) [reply src/CairoMakie.jl]: the screen is always global and toplevel, right?  The solution you posted also works, this is just an alternative.

**Pull request merged**